pkg/controller/elasticsearch/pdb/reconcile_with_roles_test.go

```diff
@@ -0,0 +1,1340 @@
+		},
+		{
+			name: "red health: 0 disruptions allowed for any role",
+			args: args{
+				es:              esv1.Elasticsearch{Status: esv1.ElasticsearchStatus{Health: esv1.ElasticsearchRedHealth}},
+				role:            []esv1.NodeRole{esv1.MasterRole, esv1.IngestRole, esv1.TransformRole, esv1.MLRole, esv1.DataFrozenRole, esv1.DataRole},
+				allStatefulSets: sset.StatefulSetList{ssetfixtures.TestSset{Replicas: 3, Master: true, Data: true}.Build()},
+			},
+			want: 0,
+		},
+		{
+			name: "green health: 1 disruption allowed for any role",
+			args: args{
+				es:              esv1.Elasticsearch{Status: esv1.ElasticsearchStatus{Health: esv1.ElasticsearchGreenHealth}},
+				role:            []esv1.NodeRole{esv1.MasterRole, esv1.IngestRole, esv1.TransformRole, esv1.MLRole, esv1.DataFrozenRole, esv1.DataRole},
+				allStatefulSets: sset.StatefulSetList{ssetfixtures.TestSset{Replicas: 3, Master: true, Data: true}.Build()},
+			},
+			want: 1,
+		},
+		{
+			name: "single-node cluster (not high-available): 1 disruption allowed for master role when the cluster is green.",
+			args: args{
+				es:              esv1.Elasticsearch{Status: esv1.ElasticsearchStatus{Health: esv1.ElasticsearchGreenHealth}},
+				role:            []esv1.NodeRole{esv1.MasterRole},
+				allStatefulSets: sset.StatefulSetList{ssetfixtures.TestSset{Replicas: 1, Master: true, Data: true}.Build()},
+			},
+			want: 1,
+		},
+		{
+			name: "single-node cluster (not high-available): 1 disruption allowed for master role when the cluster is yellow.",
+			args: args{
+				es:              esv1.Elasticsearch{Status: esv1.ElasticsearchStatus{Health: esv1.ElasticsearchYellowHealth}},
+				role:            []esv1.NodeRole{esv1.MasterRole},
+				allStatefulSets: sset.StatefulSetList{ssetfixtures.TestSset{Replicas: 1, Master: true, Data: true}.Build()},
+			},
+			want: 1,
+		},
+		{
+			name: "green health but only 1 master: 1 disruption allowed as the cluster is non-ha",
+			args: args{
+				es:   esv1.Elasticsearch{Status: esv1.ElasticsearchStatus{Health: esv1.ElasticsearchGreenHealth}},
+				role: []esv1.NodeRole{esv1.DataRole},
+				allStatefulSets: sset.StatefulSetList{
+					ssetfixtures.TestSset{Replicas: 1, Master: true, Data: false}.Build(),
+					ssetfixtures.TestSset{Replicas: 3, Master: false, Data: true}.Build(),
+					ssetfixtures.TestSset{Replicas: 2, Ingest: true}.Build(),
+				},
+			},
+			want: 1,
+		},
+		{
+			name: "green health and only 1 data node: 1 disruption allowed for data role",
+			args: args{
+				es:   esv1.Elasticsearch{Status: esv1.ElasticsearchStatus{Health: esv1.ElasticsearchGreenHealth}},
+				role: []esv1.NodeRole{esv1.DataRole},
+				allStatefulSets: sset.StatefulSetList{
+					ssetfixtures.TestSset{Replicas: 3, Master: true, Data: false}.Build(),
+					ssetfixtures.TestSset{Replicas: 1, Master: false, Data: true}.Build(),
+				},
+			},
+			want: 1,
+		},
+		{
+			name: "yellow health and only 1 data node: 0 disruption allowed for data role",
```

Comment:
Based on our discussions I would expect one disruption to be allowed as with a single node data tier there can be no expectation of HA?